src/pathfinder/water_regions.cpp

```diff
@@ -0,0 +1,405 @@
+ * @param water_region_patch Water patch within the water region to start searching from
+ * @param callback The function that will be called for each accessible water patch that is found
+ */
+void VisitWaterRegionPatchNeighbors(const WaterRegionPatchDesc& water_region_patch, TVisitWaterRegionPatchCallBack& callback)
+{
+	const WaterRegion &current_region = GetUpdatedWaterRegion(water_region_patch.x, water_region_patch.y);
+
+	/* Visit adjacent water region patches in each cardinal direction */
+	for (DiagDirection side = DIAGDIR_BEGIN; side < DIAGDIR_END; side++) VisitAdjacentWaterRegionPatchNeighbors(water_region_patch, side, callback);
+
+	/* Visit neigboring water patches accessible via cross-region aqueducts */
+	if (current_region.HasCrossRegionAqueducts()) {
+		for (const TileIndex tile : current_region) {
+			if (GetWaterRegionPatchInfo(tile) == water_region_patch && IsAqueductTile(tile)) {
+				const TileIndex other_end_tile = GetOtherBridgeEnd(tile);
+				if (GetWaterRegionIndex(tile) != GetWaterRegionIndex(other_end_tile)) callback(GetWaterRegionPatchInfo(other_end_tile));
+			}
+		}
+	}
+}
+
+void DEBUG_MarkRegion(int x, int y, int label, int color, bool override)  // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+{
+#ifndef NDEBUG
+	for (int dx = 0; dx < WATER_REGION_EDGE_LENGTH; ++dx) {
+		for (int dy = 0; dy < WATER_REGION_EDGE_LENGTH; ++dy) {
+			int tx = x * WATER_REGION_EDGE_LENGTH + dx;
+			int ty = y * WATER_REGION_EDGE_LENGTH + dy;
+			TileIndex tile = TileXY(tx, ty);
+
+			if (label < 0 || GetWaterRegionPatchInfo(tile).label == label)
+				DEBUG_addDrawRectangleInstruction(tile, color, override);  // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+		}
+	}
+#endif
+}
+
+void DEBUG_MarkRegion(TWaterRegionIndex region_index, int color, bool override)
+{
+	DEBUG_MarkRegion(GetWaterRegionX(region_index), GetWaterRegionY(region_index), -1, color, override);
+}
+
+std::vector<WaterRegionSaveLoadInfo> GetWaterRegionSaveLoadInfo()
+{
+	std::vector<WaterRegionSaveLoadInfo> result;
+	for (WaterRegion &region : _water_regions) result.push_back(WaterRegionSaveLoadInfo{ region.IsInitialized()});
+	return result;
+}
+
+void LoadWaterRegions(const std::vector<WaterRegionSaveLoadInfo> &save_load_info)
+{
+	DEBUG_clearDrawInstructions();
+	_water_regions.clear();
+	_water_regions.reserve(save_load_info.size());
+	TWaterRegionIndex index = 0;
+	for (const auto &loaded_region_info : save_load_info) {
+		const int region_x = index % GetWaterRegionMapSizeX();
+		const int region_y = index / GetWaterRegionMapSizeX();
+		WaterRegion water_region{ region_x, region_y};
+		if (loaded_region_info.initialized) water_region.ForceUpdate();
+		if (!loaded_region_info.initialized) DEBUG_MarkRegion(GetWaterRegionX(index), GetWaterRegionY(index), -1, 1); // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+		_water_regions.push_back(std::move(water_region));
```

Comment:
Maybe use the "method" of `InitializeWaterRegions`, i.e. just put the element in the `_water_regions` and then `_water_regions.back().ForceUpdate()` in the `if`-statement?